src/AppInstallerRepositoryCore/CompositeSource.cpp

```diff
@@ -0,0 +1,578 @@
+                }
+
+                switch (property)
+                {
+                case PackageProperty::Id:
+                    return truth->GetProperty(PackageVersionProperty::Id);
+                case PackageProperty::Name:
+                    return truth->GetProperty(PackageVersionProperty::Name);
+                default:
+                    THROW_HR(E_UNEXPECTED);
+                }
+            }
+
+            std::shared_ptr<IPackageVersion> GetInstalledVersion() const override
+            {
+                if (m_installedPackage)
+                {
+                    return m_installedPackage->GetInstalledVersion();
+                }
+
+                return {};
+            }
+
+            std::vector<PackageVersionKey> GetAvailableVersionKeys() const override
+            {
+                if (m_availablePackage)
+                {
+                    std::vector<PackageVersionKey> result = m_availablePackage->GetAvailableVersionKeys();
+                    std::string_view channel = m_installedChannel;
+
+                    // Remove all elements whose channel does not match the installed package.
+                    result.erase(
+                        std::remove_if(result.begin(), result.end(), [&](const PackageVersionKey& pvk) { return !Utility::CaseInsensitiveEquals(pvk.Channel, channel); }),
+                        result.end());
+
+                    return result;
+                }
+
+                return {};
+            }
+
+            std::shared_ptr<IPackageVersion> GetLatestAvailableVersion() const override
+            {
+                return GetAvailableVersion({ "", "", m_installedChannel.get() });
+            }
+
+            std::shared_ptr<IPackageVersion> GetAvailableVersion(const PackageVersionKey& versionKey) const override
+            {
+                if (m_availablePackage)
+                {
+                    return m_availablePackage->GetAvailableVersion(versionKey);
+                }
+
+                return {};
+            }
+
+            bool IsUpdateAvailable() const override
+            {
+                auto installed = GetInstalledVersion();
+
+                if (!installed)
+                {
+                    return false;
+                }
+
+                auto latest = GetLatestAvailableVersion();
+
+                return (latest && (GetVACFromVersion(installed.get()).IsUpdatedBy(GetVACFromVersion(latest.get()))));
+            }
+
+            void SetAvailablePackage(std::shared_ptr<IPackage> availablePackage)
+            {
+                m_availablePackage = std::move(availablePackage);
+            }
+
+        private:
+            std::shared_ptr<IPackage> m_installedPackage;
+            Utility::LocIndString m_installedChannel;
+            std::shared_ptr<IPackage> m_availablePackage;
+        };
+
+        // A sentinel package with an unknown version.
+        struct UnknownAvailablePackage : public IPackage

```

Comment:
>UnknownAvailablePackage [](start = 15, length = 23)

Why is this needed? Shall we just return empty if no available package matched?

Reply:
Got it, it's for ambiguous match

---
In reply to: [507624727](https://github.com/microsoft/winget-cli/pull/598#discussion_r507624727) [](ancestors = 507624727)

Reply:
Indeed; the goal is for upgrade to provide more information about the options, and once one is selected and stored in the winget installation index, to provide a better match behavior.

---
In reply to: [507650355](https://github.com/microsoft/winget-cli/pull/598#discussion_r507650355) [](ancestors = 507650355,507624727)